File: RecoverBinarySearch.java
//Approach - 1 : Inoder and Swap the nodes
//Time Complexity : O(n)
//Space Complexity : O(h)

class Solution {
    TreeNode prev;
    TreeNode first;
    TreeNode second;
    boolean flag;
    //TreeNode curr // root only
    public void recoverTree(TreeNode root) {
        if(root == null) return;
        inorder(root);
        
        //swap the values
        int temp = first.val;
        first.val = second.val;
        second.val = temp;
    }
    
    private void inorder(TreeNode root){
        //base
        if(root == null) return;
         
        //logic
        inorder(root.left);
        
        //breach happens
        if(prev != null && prev.val > root.val){
            //first breach
            if(!flag){ 
                flag = true;
                first = prev;
                second = root;
            }
            else{ // second breach
                second = root;
            }
        }
        
        prev = root;
        inorder(root.right);
  
    }
}


//Approach - 2 : Iterative by using Stack, Inoder and Swap the nodes
//Time Complexity : O(n)
//Space Complexity : O(h)

class Solution {
    TreeNode prev;
    TreeNode first;
    TreeNode second;
    boolean flag;
    //TreeNode curr // root only
    public void recoverTree(TreeNode root) {
        if(root == null) return;
        inorder(root);
        
        //swap the values
        int temp = first.val;
        first.val = second.val;
        second.val = temp;
    }
    
    private void inorder(TreeNode root){

        Stack<TreeNode> st = new Stack<>();
        while(!st.isEmpty() || root != null){
            while(root != null){
                st.push(root);
                root = root.left;
            }
        root = st.pop(); 
        
        //breach happens
        if(prev != null && prev.val > root.val){
            //first breach
            if(!flag){ 
                flag = true;
                first = prev;
                second = root;
            }
            else{ // second breach
                second = root;
            }
        }        
        prev = root;
        root = root.right;
        }
    }
}

File: recoverBinarysearch.py

# // Time Complexity :O(n)
# // Space Complexity :O(1)
# // Did this code successfully run on Leetcode :Yes
# // Any problem you faced while coding this :No
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def recoverTree(self, root: Optional[TreeNode]) -> None:
        curr=root
        first=None
        last=None
        prev=TreeNode(-sys.maxsize)
        while curr:
            if curr.left==None:
                if prev.val>curr.val:
                    if first == None:
                        first=prev
                        last=curr
                    else:
                        last=curr
                prev=curr
                curr=curr.right
            else:
                root_rightmost=curr.left
                while root_rightmost.right and root_rightmost.right!=curr:
                    root_rightmost=root_rightmost.right
                if root_rightmost.right==None:
                    root_rightmost.right=curr
                    curr=curr.left
                else:
                    if root_rightmost.right==curr:
                        root_rightmost.right=None
                        if prev.val>curr.val:
                            if first==None:
                                first=prev
                                last=curr
                            else:
                                last=curr
                        prev=curr
                        curr=curr.right

        first.val,last.val=last.val,first.val